### PR TITLE
Update for 1.17 Fabric

### PR DESCRIPTION
**Pull request by jhbuchanan45**
**TL;DR**
\> offered to get 1.16 fabric server working for friends last week
\> said I'd try 1.17
\> couldn't find 1.17 version of this mod yet
\> 'Then I'll do it myself'
\> realized I didn't know how to mod minecraft
\> learned to mod
\> seems to be working somehow

Joking aside, these should be the bulk of the changes needed for the mod to support 1.17, with the biggest chunk of work probably being rewriting how the models are created. Aside from that, the integrations with various mods should work again with their changed APIs, and I had to rewrite a bunch of stuff relating to the Trinkets mod. 
I'll hopefully be having some people give it a test run sometime in the next week, so I'll fix any errors they spot (Should only be small things now, but I haven't double-checked every single item is working properly)

The new items you added in the 1.16 version recently also still need to be ported over, and potentially re-added to the loot table (At some point I must have mistakenly pulled the WIP version with those from the 1.16 branch)

### CODE REVIEW

File: src/main/java/artifacts/trinkets/TrinketsHelper.java [hunk 1]
```diff
@@ -49,15 +51,17 @@ public static boolean isEquipped(Predicate<ItemStack> filter, LivingEntity entit
 
 		// LivingEntity not currently supported by Trinkets
```
Comment (florensie) [inline src/main/java/artifacts/trinkets/TrinketsHelper.java:52]: Trinkets 3.0 uses LivingEntity now so this check can be removed.

File: src/main/java/artifacts/mixin/mixins/client/render/PlayerEntityRendererMixin.java [hunk 2]
```diff
@@ -44,13 +47,16 @@ private static GloveItem getGloveToRenderForHand(PlayerEntity player, Hand hand)
...
-		String slotGroup = hand == Hand.MAIN_HAND ? SlotGroups.HAND : SlotGroups.OFFHAND;
-		ItemStack stack = TrinketsApi.getTrinketComponent(player).getStack(slotGroup, Slots.GLOVES);
+		String slotGroup = hand == Hand.MAIN_HAND ? "hand" : "offhand";
+		Optional<TrinketComponent> maybeComponent = TrinketsApi.getTrinketComponent(player);
 
-		if (stack.getItem() instanceof GloveItem) {
-			return (GloveItem) stack.getItem();
-		}
+		if (maybeComponent.isPresent()) {
+			ItemStack item = maybeComponent.get().getInventory().get(slotGroup).get("glove").getStack(0);
```
[the remark this answers is not in the document]
Comment (florensie) [reply src/main/java/artifacts/mixin/mixins/client/render/PlayerEntityRendererMixin.java:54]: Local variable should be called `stack` here.

File: src/main/java/artifacts/mixin/mixins/client/render/PlayerEntityRendererMixin.java [hunk 3]
```diff
@@ -44,13 +47,16 @@ private static GloveItem getGloveToRenderForHand(PlayerEntity player, Hand hand)
 			return null;
 		}
 
-		String slotGroup = hand == Hand.MAIN_HAND ? SlotGroups.HAND : SlotGroups.OFFHAND;
-		ItemStack stack = TrinketsApi.getTrinketComponent(player).getStack(slotGroup, Slots.GLOVES);
+		String slotGroup = hand == Hand.MAIN_HAND ? "hand" : "offhand";
+		Optional<TrinketComponent> maybeComponent = TrinketsApi.getTrinketComponent(player);
```
Comment (florensie) [inline src/main/java/artifacts/mixin/mixins/client/render/PlayerEntityRendererMixin.java:51]: I usually don't indicate in my variable names that they're optional with "maybe", so "comp" or "component" is fine here.

File: src/main/java/artifacts/trinkets/TrinketsHelper.java [hunk 4]
```diff
@@ -49,15 +51,17 @@ public static boolean isEquipped(Predicate<ItemStack> filter, LivingEntity entit
...
+				TrinketComponent tComponent = maybeComponent.get();
 
-				if (!stack.isEmpty() && stack.getItem() instanceof TrinketArtifactItem
-						&& (Components.ARTIFACT_ENABLED.get(stack).get() || ignoreEffectsDisabled)) {
-					stacks.add(stack);
-				}
+				tComponent.getAllEquipped().forEach(pair -> {
+					if (pair.getRight().getItem() instanceof TrinketArtifactItem &&
+							(Components.ARTIFACT_ENABLED.get(pair.getRight()).get() || ignoreEffectsDisabled)) {
+
```
Comment (florensie) [inline src/main/java/artifacts/trinkets/TrinketsHelper.java:62]: This seems broken without `stacks.add(stack)` here?

File: src/main/java/artifacts/trinkets/TrinketsHelper.java [hunk 5]
```diff
@@ -49,15 +51,17 @@ public static boolean isEquipped(Predicate<ItemStack> filter, LivingEntity entit
...
-			for (int i = 0; i < inventory.size(); i++) {
-				ItemStack stack = inventory.getStack(i);
+			if (maybeComponent.isPresent()) {
+				TrinketComponent tComponent = maybeComponent.get();
 
-				if (!stack.isEmpty() && stack.getItem() instanceof TrinketArtifactItem
-						&& (Components.ARTIFACT_ENABLED.get(stack).get() || ignoreEffectsDisabled)) {
-					stacks.add(stack);
-				}
+				tComponent.getAllEquipped().forEach(pair -> {
```
Comment (florensie) [inline src/main/java/artifacts/trinkets/TrinketsHelper.java:59]: Now that trinkets provides this getAllEquipped method we can use a stream with filters to get to our list of stacks.

File: src/main/java/artifacts/world/CampsiteFeature.java [hunk 6]
```diff
@@ -23,51 +26,65 @@
...
+	);
 
 	public CampsiteFeature() {
 		super(DefaultFeatureConfig.CODEC);
 	}
 
 	@Override
-	public boolean generate(StructureWorldAccess world, ChunkGenerator chunkGenerator, Random random, BlockPos blockPos, DefaultFeatureConfig featureConfig) {
+	public boolean generate(FeatureContext<DefaultFeatureConfig> context) {
+		BlockPos blockPos = context.getOrigin();
```
Comment (florensie) [inline src/main/java/artifacts/world/CampsiteFeature.java:84]: Feel free to use the `var` keyword for these locals, it's clear from the context which type they are.

File: src/main/java/artifacts/mixin/mixins/statuseffect/LivingEntityMixin.java [hunk 7]
```diff
@@ -30,13 +31,15 @@ public LivingEntityMixin(EntityType<?> type, World world) {
 	private void applyPermanentEffects(CallbackInfo info) {
 		if (!this.world.isClient && this.age % 15 == 0) {
 
-			TrinketsHelper.getAllEquipped((LivingEntity) (Object) this).forEach(stack -> {
+			TrinketsApi.getTrinketComponent((LivingEntity) (Object) this).ifPresent(comp -> comp.getAllEquipped().forEach(pair -> {
```
Comment (florensie) [inline src/main/java/artifacts/mixin/mixins/statuseffect/LivingEntityMixin.java:34]: This is not the same: our TrinketsHelper only returns stacks for Artifacts trinkets that are enabled, using the trinkets API you're getting all trinkets including from other mods.

File: src/main/java/artifacts/item/trinket/pendant/FlamePendantItem.java [hunk 8]
```diff
@@ -23,7 +23,6 @@ public FlamePendantItem() {
 	private static void applyEffects(LivingEntity user, Entity attacker, Random random) {
 		if (user != null && attacker != null && TrinketsHelper.isEquipped(Items.FLAME_PENDANT, user) && random.nextFloat() < 0.4f) {
 			attacker.setOnFireFor(8);
-			//noinspection ConstantConditions
```
Comment (florensie) [inline src/main/java/artifacts/item/trinket/pendant/FlamePendantItem.java:26]: Reason for removing this? Seems to me like it would still show a warning?

File: src/main/java/artifacts/item/trinket/TrinketArtifactItem.java [hunk 9]
```diff
@@ -192,32 +218,4 @@ private static String getEffectsEnabledLanguageKey(ItemStack stack) {
 		return effectsEnabled(stack) ? "artifacts.trinket.effectsenabled" : "artifacts.trinket.effectsdisabled";
 	}
 
-	// From Curios
```
Comment (florensie) [inline src/main/java/artifacts/item/trinket/TrinketArtifactItem.java:195]: You can't remove this as it serves a real purpose. Some artifacts have sounds with a different pitch an volume.

File: src/main/java/artifacts/client/render/RenderLayer.java [hunk 10]
```diff
@@ -5,26 +5,26 @@
...
 				.texture(new Texture(textureLocation, false, false))
 				.transparency(NO_TRANSPARENCY)
-				.alpha(ONE_TENTH_ALPHA)
 				.cull(DISABLE_CULLING)
 				.lightmap(ENABLE_LIGHTMAP)
 				.overlay(ENABLE_OVERLAY_COLOR)
 				.build(true);
-		return of("artifacts_entity_unlit", VertexFormats.POSITION_COLOR_TEXTURE_OVERLAY_LIGHT_NORMAL, GL11.GL_QUADS, 256, true, false, renderState);
+		// TODO - Find cleaner solution that doesn't use mixins
+		return RenderLayer.of("artifacts_entity_unlit", VertexFormats.POSITION_COLOR_TEXTURE_OVERLAY_LIGHT_NORMAL, VertexFormat.DrawMode.QUADS, 256, true, false, renderState);
```
Comment (florensie) [inline src/main/java/artifacts/client/render/RenderLayer.java:28]:
```suggestion
		return of("artifacts_entity_unlit", VertexFormats.POSITION_COLOR_TEXTURE_OVERLAY_LIGHT_NORMAL, VertexFormat.DrawMode.QUADS, 256, true, false, renderState);
```
This might work too.

**Review by florensie — COMMENTED** [review]
I got half-way through reviewing this but I don't think I'll be able to merge this. I had planned to first catch up with the updates on 1.16 before I update to 1.17. If I merge this and then finish updating 1.16 it would be one hell of a merge conflict to get the new stuff on the 1.17 branch.

I really appreciate the effort you put into this but it's often important that you ask before contributing so you don't waste your time. That said, your PR has some really helpful things thought out in regards to the update and I'm sure it will be a great reference when I get to it, hopefully this week.

**Comment by jhbuchanan45** [comment]
Thanks a ton for taking the time to review even part of this, and frankly I totally agree with your sentiment, especially having flicked through your review and seen some of the bizarre decisions I made. It would definitely make sense to start from scratch, given you know the new trinkets API (and how the old one was used) much better than I managed to guess of it in a few days. Frankly I'm still surprised my version even runs xD

As I said, this whole thing started as a bit of a joke to myself to see if I could actually get it going, so the experience of getting this far was enough to make it worth my time, and I'm glad it can at least serve as a reference (those models in particular were a pain to figure out the first time, and were a bunch of pretty tedious of copy pasting)

I look forwards to the official, much better version coming out.
